sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/state/MeterSharedState.java

```diff
@@ -102,48 +99,43 @@ public final void registerLongAsynchronousInstrument(
-    // TODO - we should avoid registering independent storage that calls observables over and over.
-    List<View> views =
-        meterProviderSharedState
-            .getViewRegistry()
-            .findViews(instrument, getInstrumentationLibraryInfo());
-    for (View view : views) {
-      MetricStorage currentStorage =
-          AsynchronousMetricStorage.longAsynchronousAccumulator(view, instrument, metricUpdater);
-      // TODO - move this in a better location.
-      if (AsynchronousMetricStorage.empty().equals(currentStorage)) {
-        continue;
-      }
-      try {
-        getMetricStorageRegistry().register(currentStorage);
-      } catch (DuplicateMetricStorageException e) {
-        logger.log(Level.WARNING, e, () -> "Failed to register metric.");
-      }
-    }
+
+    meterProviderSharedState
+        .getViewRegistry()
+        .findViews(instrument, getInstrumentationLibraryInfo())
+        .stream()
+        .map(
+            view ->
+                AsynchronousMetricStorage.longAsynchronousAccumulator(
+                    view, instrument, metricUpdater))
+        .filter(m -> !m.isEmpty())
+        .forEach(this::register);
   }
 
   /** Registers new asynchronous storage associated with a given {@code double} instrument. */
   public final void registerDoubleAsynchronousInstrument(
       InstrumentDescriptor instrument,
       MeterProviderSharedState meterProviderSharedState,
       Consumer<ObservableDoubleMeasurement> metricUpdater) {
-    // TODO - we should avoid registering independent storage that calls observables over and over.
-    List<View> views =
-        meterProviderSharedState
-            .getViewRegistry()
-            .findViews(instrument, getInstrumentationLibraryInfo());
-    for (View view : views) {
-      MetricStorage currentStorage =
-          AsynchronousMetricStorage.doubleAsynchronousAccumulator(view, instrument, metricUpdater);
-      // TODO - move this in a better location.
-      if (AsynchronousMetricStorage.empty() == currentStorage) {
-        continue;
-      }
-      try {
-        getMetricStorageRegistry().register(currentStorage);
-      } catch (DuplicateMetricStorageException e) {
-        logger.log(Level.WARNING, e, () -> "Failed to register metric.");
-      }
+
+    meterProviderSharedState
+        .getViewRegistry()
+        .findViews(instrument, getInstrumentationLibraryInfo())
+        .stream()
+        .map(
+            view ->
+                AsynchronousMetricStorage.doubleAsynchronousAccumulator(
+                    view, instrument, metricUpdater))
+        .filter(m -> !m.isEmpty())
+        .forEach(this::register);
+  }
+
+  private <S extends MetricStorage> S register(S storage) {
+    try {
+      return getMetricStorageRegistry().register(storage);
+    } catch (DuplicateMetricStorageException e) {
+      logger.log(Level.WARNING, e, () -> DebugUtils.duplicateMetricErrorMessage(e));
     }
+    return null;
```

Comment:
Nit: As we're moving to non-null here, might make sense to use a "NoMetricStorage", or alter the return to be Optional<S>

Reply:
We avoid `Optional` in this codebase (it's equivalent to Nullable given our checker setup) but an empty type would be good if possible

Reply:
It's a private stealth method. Having it void would be the best, but I need the returned value in the first use case (could iterate again to register though).

Reply:
Think this method needs the Nullable annotation - I merged in main to the PR to have the nullness checking activated

Reply:
Added.